aws/rust-runtime/aws-inlineable/src/s3_express.rs

```diff
@@ -93,23 +93,398 @@ pub(crate) mod auth {
+            let (size, capacity) = {
+                let cache = self.inner.lock().unwrap();
+                (cache.len(), cache.cap())
+            };
+            write!(
+                f,
+                "S3ExpressIdentityCache {{ time_source: {:?}, buffer_time: {:?} }}, with size/capacity: {}/{}",
+                self.time_source, &self.buffer_time, size, capacity,
+            )
+        }
+    }
+
+    impl S3ExpressIdentityCache {
+        pub(crate) fn new(
+            capacity: usize,
+            time_source: SharedTimeSource,
+            buffer_time: Duration,
+        ) -> Self {
+            let mut rng = Rng::default();
+            let mut random_bytes = [0u8; 64];
+            rng.fill(&mut random_bytes);
+            Self {
+                inner: Mutex::new(LruCache::new(NonZeroUsize::new(capacity).unwrap())),
+                time_source,
+                buffer_time,
+                random_bytes,
+            }
+        }
+
+        pub(crate) fn key(&self, bucket_name: &str, creds: &Credentials) -> CacheKey {
+            CacheKey({
+                let mut mac = Hmac::<Sha256>::new_from_slice(self.random_bytes.as_slice())
+                    .expect("should be created from random 64 bytes");
+                let input = format!("{}{}", creds.access_key_id(), creds.secret_access_key());
+                mac.update(input.as_ref());
+                let mut inner = hex::encode(mac.finalize_fixed());
+                inner.push_str(bucket_name);
+                inner
+            })
+        }
+
+        pub(crate) async fn get_or_load<F, Fut>(
+            &self,
+            key: CacheKey,
+            loader: F,
+        ) -> Result<Identity, BoxError>
+        where
+            F: FnOnce() -> Fut,
+            Fut: Future<Output = Result<(Identity, SystemTime), BoxError>>,
+        {
+            let expiring_cache = {
+                let mut inner = self.inner.lock().unwrap();
+                inner
+                    .get_or_insert_mut(key, || ExpiringCache::new(self.buffer_time))
+                    .clone()
+            };
+
+            let now = self.time_source.now();
+
+            match expiring_cache.yield_or_clear_if_expired(now).await {
+                Some(identity) => {
+                    tracing::debug!(
+                        buffer_time=?self.buffer_time,
+                        cached_expiration=?identity.expiration(),
+                        now=?now,
+                        "loaded identity from cache"
+                    );
+                    Ok(identity)
+                }
+                None => {
+                    let start_time = self.time_source.now();
+                    let identity = expiring_cache.get_or_load(loader).await;
+                    tracing::info!(
+                        "identity cache miss occurred; added new identity (took {:?})",
+                        self.time_source.now().duration_since(start_time)
+                    );
```

Comment:
Consider porting over @rcoh's changes in #3428

Reply:
Thought about the same while reviewing the above PR :) Will do.

Reply:
Addressed in f2fbae9.